plugins/cassandra/cassandraembed.cpp

```diff
@@ -3695,21 +3796,127 @@ class CassSortableIterator : public CassandraIterator
     bool descending;
 };
 
-class CassMultiIterator : public CInterface, implements IRowProvider, implements ICompare, implements IConstWorkUnitIterator
+interface IConstWorkUnitIteratorEx : public IConstWorkUnitIterator
+{
+    virtual bool hasPostFilters() const = 0;
+    virtual bool isMerging() const = 0;
+};
+
+/*
+ *
+ * The cache entries serve two purposes:
+ *
+ * 1. They allow us to map row numbers to values for the end of each page returned, which can make forward paging efficient when not post-sorting
+ * 2. They allow us to preserve post-sort results in order to avoid having to re-retrieve them.
+ */
+
+class CCassandraWuUQueryCacheEntry : public CInterfaceOf<IInterface>
+{
+public:
+    CCassandraWuUQueryCacheEntry()
+    {
+        hint = get_cycles_now(); // MORE - should do better perhaps?
+        lastAccess = msTick();
+    }
+    __int64 queryHint() const
+    {
+        return hint;
+    }
+    void noteWuid(const char *wuid, const char *fieldValue, unsigned row)
+    {
+        CriticalBlock b(crit);
+        ForEachItemInRev(idx, rows)
```

Comment:
How big is this going to get?  Inserting could get painful if there are large numbers of following rows.